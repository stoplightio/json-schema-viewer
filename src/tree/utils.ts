import { isPlainObject } from '@stoplight/json';
import {
  type MirroredSchemaNode,
  type ReferenceNode,
  type RegularNode,
  type SchemaNode,
  isBooleanishNode,
  isReferenceNode,
  isRegularNode,
  isRootNode,
  SchemaNodeKind,
} from '@stoplight/json-schema-tree';
import { BooleanishNode } from '@stoplight/json-schema-tree/nodes/BooleanishNode';

import { isNonNullable } from '../guards/isNonNullable';
import type { ViewMode } from '../types';
import type {
  ArrayNode,
  ComplexArrayNode,
  ComplexDictionaryNode,
  DictionaryNode,
  FlattenableNode,
  PrimitiveArrayNode,
  PrimitiveDictionaryNode,
} from './types';

export type ChildNode = RegularNode | ReferenceNode | MirroredSchemaNode;

export const isNonEmptyParentNode = (
  node: SchemaNode,
): node is RegularNode & {
  children: ChildNode[] & { 0: ChildNode };
} => isRegularNode(node) && !!node.children && node.children.length > 0;

export function isFlattenableNode(node: SchemaNode): node is FlattenableNode {
  if (!isRegularNode(node)) return false;

  if ((!isArrayNode(node) && !isDictionaryNode(node)) || !isNonNullable(node.children) || node.children.length === 0) {
    return false;
  }

  return (
    node.children.length === 1 &&
    ((isRegularNode(node.children[0]) && (!isArrayNode(node) || !isDictionaryNode(node.children[0]))) ||
      (isReferenceNode(node.children[0]) && node.children[0].error !== null))
  );
}

export function isPrimitiveArray(node: SchemaNode): node is PrimitiveArrayNode {
  return isFlattenableNode(node) && isArrayNode(node) && isRegularNode(node.children[0]) && node.children[0].simple;
}

export function isPrimitiveDictionary(node: SchemaNode): node is PrimitiveDictionaryNode {
  return (
    isFlattenableNode(node) && isDictionaryNode(node) && isRegularNode(node.children[0]) && node.children[0].simple
  );
}

export function isComplexArray(node: SchemaNode): node is ComplexArrayNode {
  return isFlattenableNode(node) && isArrayNode(node) && isRegularNode(node.children[0]) && !node.children[0].simple;
}

export function isComplexDictionary(node: SchemaNode): node is ComplexDictionaryNode {
  return (
    isFlattenableNode(node) && isDictionaryNode(node) && isRegularNode(node.children[0]) && !node.children[0].simple
  );
}

export function isDictionaryNode(node: SchemaNode): node is DictionaryNode {
  return (
    isRegularNode(node) &&
    node.primaryType === SchemaNodeKind.Object &&
    isPlainObject(node.fragment.additionalProperties)
  );
}

export function isArrayNode(node: SchemaNode): node is ArrayNode {
  return isRegularNode(node) && node.primaryType === SchemaNodeKind.Array;
}

/**
 * Returns the children of `node` that should be displayed in a viewer or
 * editor. Defaults to `node.children`, except for Arrays that get special
 * handling (flattening).
 */
export function visibleChildren(node: SchemaNode): SchemaNode[] {
  if (!isRegularNode(node) || isPrimitiveArray(node) || isPrimitiveDictionary(node)) {
    return [];
  }
  if (isComplexArray(node) || isComplexDictionary(node)) {
    // flatten the tree here, and show the properties of the item type directly
    return node.children[0].children ?? [];
  }
  return node.children ?? [];
}

export function isPropertyRequired(schemaNode: SchemaNode): boolean {
  const { parent } = schemaNode;
  if (parent === null || !isRegularNode(parent) || schemaNode.subpath.length === 0) {
    return false;
  }

  return !!parent.required?.includes(schemaNode.subpath[schemaNode.subpath.length - 1]);
}

function isValidViewMode(node: RegularNode, viewMode: ViewMode): boolean {
  const { validations } = node;

  if (!!validations.writeOnly === !!validations.readOnly) {
    return true;
  }

  return !((viewMode === 'read' && !!validations.writeOnly) || (viewMode === 'write' && !!validations.readOnly));
}

function isRenderableNode(node: BooleanishNode | RegularNode): boolean {
  if (node.parent === null) return true;

  if (isDictionaryNode(node.parent)) {
    // if dictionary, do not render explicitly defined properties
    return node.subpath.length !== 2 || node.subpath[0] !== 'properties';
  }

  // do not render additionalItems
  if (isArrayNode(node.parent)) {
    return node.subpath[0] !== 'additionalItems';
  }

  // do not render true/false additionalProperties
  if (isRegularNode(node.parent) && node.parent.primaryType === SchemaNodeKind.Object && isBooleanishNode(node)) {
    return !(node.subpath.length === 1 || node.subpath[0] === 'additionalProperties');
  }

  return true;
}

export function shouldNodeBeIncluded(node: SchemaNode, viewMode: ViewMode = 'standalone'): boolean {
  return (
    (isReferenceNode(node) || isRootNode(node) || isRenderableNode(node)) &&
    (!isRegularNode(node) || isValidViewMode(node, viewMode))
  );
}
