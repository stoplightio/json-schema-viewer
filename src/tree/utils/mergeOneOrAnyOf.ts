import { JSONSchema4 } from 'json-schema';

export function mergeOneOrAnyOf(schema: JSONSchema4, combiner: 'oneOf' | 'anyOf'): JSONSchema4[] {
  const merged: JSONSchema4[] = [];

  if (!Array.isArray(schema.allOf)) return merged;

  const items = schema[combiner];

  if (Array.isArray(items)) {
    for (const item of items) {
      merged.push({
        allOf: [item, ...schema.allOf],
      });
    }
  }

  return merged;
}
