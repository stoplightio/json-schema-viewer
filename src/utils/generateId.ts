export const generateId = (): string => Math.random().toString(36);
