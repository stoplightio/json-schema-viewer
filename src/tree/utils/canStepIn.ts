import { JSONSchema4 } from 'json-schema';

import { SchemaKind } from '../../types';
import { getCombiners } from '../../utils/getCombiners';
import { getPrimaryType } from '../../utils/getPrimaryType';

export const canStepIn = (fragment: JSONSchema4) => {
  if (getCombiners(fragment) !== void 0) {
    return true;
  }

  const type = getPrimaryType(fragment);
  return type === SchemaKind.Array || type === SchemaKind.Object;
};
