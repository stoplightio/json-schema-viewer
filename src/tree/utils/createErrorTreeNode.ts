import { TreeListParentNode } from '@stoplight/tree-list';

import { SchemaTreeListNode } from '../../types';
import { generateId } from '../../utils/generateId';
import { getSchemaNodeMetadata, metadataStore } from '../metadata';

export function createErrorTreeNode(parent: TreeListParentNode, error: string) {
  const { path } = getSchemaNodeMetadata(parent);

  const errorNode: SchemaTreeListNode = {
    id: generateId(),
    name: '',
    parent,
  };

  metadataStore.set(errorNode, {
    path,
    error,
  });

  return errorNode;
}
