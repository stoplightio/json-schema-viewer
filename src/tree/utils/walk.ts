import { Optional } from '@stoplight/types';
import { JSONSchema4 } from 'json-schema';
import { isObject as _isObject } from 'lodash';

import { IArrayNode, IBaseNode, ICombinerNode, IObjectNode, SchemaKind, SchemaNode } from '../../types';
import { flattenTypes } from '../../utils/flattenTypes';
import { generateId } from '../../utils/generateId';
import { getAnnotations } from '../../utils/getAnnotations';
import { getCombiners } from '../../utils/getCombiners';
import { getPrimaryType } from '../../utils/getPrimaryType';
import { getValidations } from '../../utils/getValidations';
import { inferType } from '../../utils/inferType';
import { normalizeRequired } from '../../utils/normalizeRequired';

function assignNodeSpecificFields(base: IBaseNode, node: JSONSchema4) {
  switch (getPrimaryType(node)) {
    case SchemaKind.Array:
      (base as IArrayNode).items = unwrapItemsOrUndefined(node.items);
      (base as IArrayNode).additionalItems =
        typeof node.additionalItems === 'boolean' ? node.additionalItems : unwrapItemsOrUndefined(node.additionalItems);
      break;
    case SchemaKind.Object:
      (base as IObjectNode).properties = unwrapPropertiesOrUndefined(node.properties);
      (base as IObjectNode).patternProperties = unwrapPropertiesOrUndefined(node.patternProperties);
      (base as IObjectNode).additionalProperties =
        typeof node.additionalProperties === 'boolean'
          ? node.additionalProperties
          : unwrapPropertiesOrUndefined(node.additionalProperties);
      break;
  }
}

export function* processNode(node: JSONSchema4): IterableIterator<SchemaNode> {
  const combiners = getCombiners(node);
  const type = node.type ?? inferType(node);
  const title = typeof node.title === 'string' ? { title: node.title } : null;

  if (combiners !== void 0) {
    for (const combiner of combiners) {
      const properties = node[combiner];
      const combinerNode: ICombinerNode = {
        id: generateId(),
        combiner,
        properties: Array.isArray(properties) ? properties.slice() : properties,
        annotations: getAnnotations(node),
        ...(type !== void 0 && { type: flattenTypes(type) }),
        ...title,
      };

      yield combinerNode;
    }
  } else if (type) {
    const base: IBaseNode = {
      id: generateId(),
      type: flattenTypes(type),
      validations: getValidations(node),
      annotations: getAnnotations(node),
      ...('required' in node && { required: normalizeRequired(node.required) }),
      enum: node.enum,
      ...title,
    };

    assignNodeSpecificFields(base, node);

    yield base;
  } else if ('enum' in node) {
    yield {
      id: generateId(),
      validations: getValidations(node),
      annotations: getAnnotations(node),
      enum: node.enum,
      ...title,
    };
  } else if ('$ref' in node) {
    yield {
      id: generateId(),
      $ref: typeof node.$ref !== 'string' ? null : node.$ref,
      ...title,
    };
  } else if ('not' in node) {
    // todo: shall we support it?
  } else {
    yield {
      id: generateId(),
      validations: {},
      annotations: {},
    };
  }
}

export type WalkerValue = {
  node: SchemaNode;
  fragment: JSONSchema4;
};

export function* walk(schema: JSONSchema4[] | JSONSchema4): IterableIterator<WalkerValue> {
  if (Array.isArray(schema)) {
    for (const segment of schema) {
      yield* walk(segment);
    }
  } else {
    for (const node of processNode(schema)) {
      yield {
        node,
        fragment: schema,
      };
    }
  }
}

function unwrapItemsOrUndefined<T = unknown>(value: T): Optional<T> {
  return _isObject(value) ? value : void 0;
}

function unwrapPropertiesOrUndefined<T = unknown>(value: T): Optional<T> {
  return _isObject(value) && !Array.isArray(value) ? value : void 0;
}
