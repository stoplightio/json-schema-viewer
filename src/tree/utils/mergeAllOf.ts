import { pathToPointer, safeStringify } from '@stoplight/json';
import { JsonPath } from '@stoplight/types';
import { JSONSchema4 } from 'json-schema';
import { ResolvingError } from '../../errors';
import { WalkingOptions } from './populateTree';

const resolveAllOf = require('@stoplight/json-schema-merge-allof');

const store = new WeakMap<WalkingOptions, WeakMap<JSONSchema4, string[]>>();

function _mergeAllOf(schema: JSONSchema4, path: JsonPath, opts: WalkingOptions) {
  return resolveAllOf(schema, {
    deep: false,
    resolvers: resolveAllOf.stoplightResolvers,
    $refResolver($ref: unknown) {
      if (typeof $ref !== 'string') {
        return {};
      }

      if (pathToPointer(path).startsWith($ref)) {
        throw new ResolvingError('Circular reference detected');
      }

      const allRefs = store.get(opts)!;
      let schemaRefs = allRefs.get(schema);

      if (schemaRefs === void 0) {
        schemaRefs = [$ref];
        allRefs.set(schema, schemaRefs);
      } else if (schemaRefs.includes($ref)) {
        const safelyResolved = JSON.parse(safeStringify(opts.resolveRef(null, $ref)));
        return 'allOf' in safelyResolved ? _mergeAllOf(safelyResolved, path, opts) : safelyResolved;
      } else {
        schemaRefs.push($ref);
      }

      const resolved = opts.resolveRef(null, $ref);

      if (Array.isArray(resolved.allOf)) {
        for (const member of resolved.allOf) {
          if (typeof member.$ref === 'string' && schemaRefs.includes(member.$ref)) {
            throw new ResolvingError('Circular reference detected');
          }
        }
      }

      return resolved;
    },
  });
}

export const mergeAllOf = (schema: JSONSchema4, path: JsonPath, opts: WalkingOptions) => {
  try {
    if (!store.has(opts)) {
      store.set(opts, new WeakMap());
    }

    return _mergeAllOf(schema, path, opts);
  } catch (ex) {
    console.info(ex.message);
    throw ex;
  }
};
