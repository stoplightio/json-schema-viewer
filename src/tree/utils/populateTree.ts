import { TreeListNode, TreeListParentNode } from '@stoplight/tree-list';
import { JsonPath, Optional } from '@stoplight/types';
import { JSONSchema4 } from 'json-schema';
import { isObject as _isObject } from 'lodash';
import { IArrayNode, IObjectNode, SchemaKind, SchemaNode, SchemaTreeListNode } from '../../types';
import { generateId } from '../../utils/generateId';
import { getCombiner } from '../../utils/getCombiner';
import { getPrimaryType } from '../../utils/getPrimaryType';
import { isCombinerNode, isRefNode } from '../../utils/guards';
import { getNodeMetadata, getSchemaNodeMetadata, metadataStore } from '../metadata';
import { createErrorTreeNode } from './createErrorTreeNode';
import { mergeAllOf } from './mergeAllOf';
import { processNode, walk } from './walk';

export type WalkerRefResolver = (path: JsonPath | null, $ref: string) => JSONSchema4;

export type WalkingOptions = {
  mergeAllOf: boolean;
  onNode?(fragment: JSONSchema4, node: SchemaNode, parentTreeNode: TreeListNode, level: number): boolean | void;
  stepIn?: boolean;
  resolveRef: WalkerRefResolver;
  shouldResolveEagerly: boolean;
};

export type Walker = (
  schema: Optional<JSONSchema4 | null>,
  parent: TreeListParentNode,
  level: number,
  path: JsonPath,
  options: WalkingOptions | null,
) => undefined;

export const populateTree: Walker = (schema, parent, level, path, options): undefined => {
  const actualSchema = prepareSchema(schema, parent, path, options);

  if (!_isObject(actualSchema)) return;

  for (const { node, fragment } of walk(actualSchema)) {
    if (options !== null && options.onNode !== void 0 && !options.onNode(fragment, node, parent, level)) continue;

    const treeNode: SchemaTreeListNode = {
      id: node.id,
      name: '',
      parent,
    };

    parent.children.push(treeNode);
    metadataStore.set(treeNode, {
      schemaNode: node,
      schema: fragment,
      path,
    });

    if (isRefNode(node) && node.$ref !== null) {
      processRef(treeNode, node as JSONSchema4, level, path, options);
    } else if (!isCombinerNode(node)) {
      switch (getPrimaryType(node)) {
        case SchemaKind.Array:
          processArray(treeNode, node as IArrayNode, level, path, options);
          break;
        case SchemaKind.Object:
          processObject(treeNode, node as IObjectNode, level, path, options);
          break;
      }
    } else if (node.combiner === 'allOf' && options?.mergeAllOf) {
      try {
        const merged = mergeAllOf(fragment, path, options);
        parent.children.pop();
        populateTree(merged, parent, level, path, options);
      } catch (ex) {
        if (Array.isArray(fragment.allOf)) {
          (treeNode as TreeListParentNode).children = [];
          bailAllOf(treeNode as TreeListParentNode, fragment, level + 1, [...path, 'allOf'], options);
        }
      }
    } else if (_isObject(node.properties)) {
      (treeNode as TreeListParentNode).children = [];

      for (const [i, property] of node.properties.entries()) {
        if ('type' in node) {
          node.properties[i] = {
            ...property,
            type: property.type || node.type,
          };
        }

        populateTree(
          node.properties[i],
          treeNode as TreeListParentNode,
          level + 1,
          [...path, node.combiner, i],
          options,
        );
      }
    }
  }

  return;
};

function processArray(
  node: SchemaTreeListNode,
  schema: IArrayNode,
  level: number,
  path: JsonPath,
  options: WalkingOptions | null,
): SchemaTreeListNode {
  const items = prepareSchema(schema.items, node, path, options);

  if (!_isObject(items)) return node;

  if (items !== schema.items) {
    // we need to update the stored metadata to make sure the subtype of given array correctly inferred by Property component
    const metadata = getNodeMetadata(node);
    if ('schemaNode' in metadata) {
      (metadata.schemaNode as IArrayNode).items = items;
    }
  }

  if ('$ref' in items) {
    processRef(node, items, level, path, options);
  } else if (Array.isArray(items)) {
    const children: SchemaTreeListNode[] = [];
    (node as TreeListParentNode).children = children;
    for (const [i, property] of items.entries()) {
      const child = populateTree(property, node as TreeListParentNode, level + 1, [...path, 'items', i], options);
      if (child !== void 0) {
        children.push(child);
      }
    }
  } else {
    const subtype = getPrimaryType(items);
    switch (subtype) {
      case SchemaKind.Object:
        return processObject(node, items as IObjectNode, level, [...path, 'items'], options);
      case SchemaKind.Array:
        return processArray(node, items as IObjectNode, level, [...path, 'items'], options);
      default:
        const combiner = getCombiner(items);
        if (combiner) {
          (node as TreeListParentNode).children = [];
          populateTree(items, node as TreeListParentNode, level, [...path, 'items'], options);
        }
    }
  }

  return node;
}

function processObject(
  node: TreeListNode,
  schema: IObjectNode,
  level: number,
  path: JsonPath,
  options: WalkingOptions | null,
): TreeListNode {
  const children: TreeListNode[] = [];

  if (_isObject(schema.properties)) {
    (node as TreeListParentNode).children = children;

    for (const [prop, property] of Object.entries(schema.properties)) {
      const child = populateTree(
        property,
        node as TreeListParentNode,
        level + 1,
        [...path, 'properties', prop],
        options,
      );
      if (child !== void 0) {
        children.push(child);
      }
    }
  }

  if (_isObject(schema.patternProperties)) {
    (node as TreeListParentNode).children = children;

    for (const [prop, property] of Object.entries(schema.patternProperties)) {
      const child = populateTree(
        property,
        node as TreeListParentNode,
        level + 1,
        [...path, 'patternProperties', prop],
        options,
      );
      if (child !== void 0) {
        children.push(child);
      }
    }
  }

  return node;
}

function processRef(
  node: TreeListNode,
  schema: JSONSchema4,
  level: number,
  path: JsonPath,
  options: WalkingOptions | null,
) {
  (node as TreeListParentNode).children = [];
  try {
    const resolved = resolveSchema(schema, path, options);
    if (_isObject(resolved) && typeof resolved.title === 'string') {
      const { schemaNode } = getSchemaNodeMetadata(node);
      schemaNode.title = resolved.title;
    }
  } catch {
    // resolving failed, nothing bad. We just won't have the title
  }
}

function bailAllOf(
  node: TreeListParentNode,
  schema: JSONSchema4,
  level: number,
  path: JsonPath,
  options: WalkingOptions | null,
) {
  if (Array.isArray(schema.allOf)) {
    for (const [i, item] of schema.allOf.entries()) {
      populateTree(item, node, level, [...path, i], options);
    }
  }
}

function resolveSchema(schema: Optional<JSONSchema4 | null>, path: JsonPath, options: WalkingOptions | null) {
  if (!_isObject(schema) || options === null || !('$ref' in schema) || typeof schema.$ref !== 'string') {
    return schema;
  }

  const resolved = options.resolveRef(path, schema.$ref);
  return _isObject(resolved) ? resolved : schema;
}

function prepareSchema(
  schema: Optional<JSONSchema4 | null>,
  node: TreeListNode,
  path: JsonPath,
  options: WalkingOptions | null,
): Optional<JSONSchema4 | null> {
  if (options === null || !options.shouldResolveEagerly) return schema;

  try {
    return resolveSchema(schema, path, options);
  } catch (ex) {
    const treeNode: SchemaTreeListNode = {
      id: generateId(),
      name: '',
      parent: node as TreeListParentNode,
      children: [],
    };

    (node as TreeListParentNode).children.push(treeNode);
    metadataStore.set(treeNode, {
      schemaNode: processNode(schema || {}),
      schema: schema || {},
      path,
    });

    treeNode.children.push(createErrorTreeNode(treeNode, ex.message));

    return null;
  }
}
