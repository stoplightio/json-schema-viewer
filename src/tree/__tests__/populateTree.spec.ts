import { Tree } from '@stoplight/tree-list';
import * as fs from 'fs';
import * as path from 'path';
import { generateId } from '../../utils/generateId';
import { populateTree } from '../populateTree';

const BASE_PATH = path.resolve(__dirname, '../../__fixtures__/');

jest.mock('../../utils/generateId', () => ({
  generateId: jest.fn(() => 'random-id'),
}));

describe('populateTree util', () => {
  it.each([
    'default-schema.json',
    'ref/original.json',
    'combiner-schema.json',
    'array-of-objects.json',
    'array-of-refs.json',
    'array-of-allofs.json',
    'todo-allof.schema.json',
    'tickets.schema.json',
    'nullish-ref.schema.json',
  ])('should match %s', filename => {
    const schema = JSON.parse(fs.readFileSync(path.resolve(BASE_PATH, filename), 'utf8'));
    const root = Tree.createArtificialRoot();
    root.id = generateId();
    populateTree(schema, root, 0, [], null);
    expect(root).toMatchSnapshot();
  });

  it.each([
    'default-schema.json',
    'ref/original.json',
    'combiner-schema.json',
    'array-of-objects.json',
    'array-of-refs.json',
    'array-of-allofs.json',
    'tickets.schema.json',
  ])('should not mutate original object %s', filename => {
    const content = fs.readFileSync(path.resolve(BASE_PATH, filename), 'utf8');
    const schema = JSON.parse(content);

    const root = Tree.createArtificialRoot();
    populateTree(schema, root, 0, [], null);
    expect(schema).toStrictEqual(JSON.parse(content));
  });

  it('given schema with complex types, throws', () => {
    const schema = {
      type: [
        'null',
        {
          type: 'object',
          properties: {
            taskId: {
              type: 'string',
              format: 'uuid',
            },
          },
        },
      ],
    };

    const root = Tree.createArtificialRoot();
    expect(() => populateTree(schema as any, root, 0, [], null)).toThrow(
      'The "type" property must be a string, or an array of strings. Objects and array of objects are not valid.',
    );
  });
});
