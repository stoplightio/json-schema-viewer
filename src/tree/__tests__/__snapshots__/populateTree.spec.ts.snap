// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populateTree util should match array-of-allofs.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
      ],
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match array-of-objects.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
          ],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
      ],
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match array-of-refs.json 1`] = `
Object {
  "children": Array [
    Object {
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match combiner-schema.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
          ],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "children": Array [
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
          ],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
      ],
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match default-schema.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
              ],
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
          ],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "children": Array [
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
          ],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "children": Array [],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
      ],
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match nullish-ref.schema.json 1`] = `
Object {
  "children": Array [
    Object {
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match ref/original.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "children": Array [],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
      ],
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match tickets.schema.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
        Object {
          "children": Array [
            Object {
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
            Object {
              "children": Array [
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
                Object {
                  "children": Array [
                    Object {
                      "id": "random-id",
                      "name": "",
                      "parent": [Circular],
                    },
                    Object {
                      "id": "random-id",
                      "name": "",
                      "parent": [Circular],
                    },
                    Object {
                      "id": "random-id",
                      "name": "",
                      "parent": [Circular],
                    },
                    Object {
                      "children": Array [
                        Object {
                          "id": "random-id",
                          "name": "",
                          "parent": [Circular],
                        },
                        Object {
                          "id": "random-id",
                          "name": "",
                          "parent": [Circular],
                        },
                        Object {
                          "id": "random-id",
                          "name": "",
                          "parent": [Circular],
                        },
                        Object {
                          "id": "random-id",
                          "name": "",
                          "parent": [Circular],
                        },
                        Object {
                          "id": "random-id",
                          "name": "",
                          "parent": [Circular],
                        },
                      ],
                      "id": "random-id",
                      "name": "",
                      "parent": [Circular],
                    },
                  ],
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
              ],
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
          ],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
      ],
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;

exports[`populateTree util should match todo-allof.schema.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
              ],
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
            Object {
              "children": Array [
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
                Object {
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
                Object {
                  "children": Array [
                    Object {
                      "id": "random-id",
                      "name": "",
                      "parent": [Circular],
                    },
                    Object {
                      "id": "random-id",
                      "name": "",
                      "parent": [Circular],
                    },
                  ],
                  "id": "random-id",
                  "name": "",
                  "parent": [Circular],
                },
              ],
              "id": "random-id",
              "name": "",
              "parent": [Circular],
            },
          ],
          "id": "random-id",
          "name": "",
          "parent": [Circular],
        },
      ],
      "id": "random-id",
      "name": "",
      "parent": [Circular],
    },
  ],
  "id": "random-id",
  "name": "",
  "parent": null,
}
`;
